crates/ethereum-forks/src/hardforks/mod.rs

```diff
@@ -0,0 +1,113 @@
+/// Ethereum helper methods
+mod ethereum;
+pub use ethereum::{EthereumActivations, EthereumHardforks};
+
+/// Optimism helper methods
+#[cfg(feature = "optimism")]
+mod optimism;
+#[cfg(feature = "optimism")]
+pub use optimism::{OptimismActivations, OptimismHardforks};
+
+use crate::{ForkCondition, Hardfork};
+use core::ops::Deref;
+
+/// Generic trait over a set of ordered hardforks
+pub trait Hardforks: Default + Clone {
+    /// Retrieves [`ForkCondition`] from `fork`. If `fork` is not present, returns
+    /// [`ForkCondition::Never`].
+    fn fork<H: Hardfork>(&self, fork: H) -> ForkCondition;
+
+    /// Get an iterator of all hardforks with their respective activation conditions.
+    fn forks_iter(&self) -> impl Iterator<Item = (&Box<dyn Hardfork + 'static>, ForkCondition)>;
+
+    /// Convenience method to check if a fork is active at a given timestamp.
+    fn is_fork_active_at_timestamp<H: Hardfork>(&self, fork: H, timestamp: u64) -> bool {
+        self.fork(fork).active_at_timestamp(timestamp)
+    }
+
+    /// Convenience method to check if a fork is active at a given block number.
+    fn is_fork_active_at_block<H: Hardfork>(&self, fork: H, block_number: u64) -> bool {
+        self.fork(fork).active_at_block(block_number)
+    }
+}
+
+/// Ordered list of a chain hardforks that implement [`Hardfork`].
+#[derive(Default, Clone, PartialEq, Eq)]
+pub struct ChainHardforks(pub Vec<(Box<dyn Hardfork>, ForkCondition)>);
```

Comment:
this is now basically the `Other(String)` approach, but matching is now entirely based on comparing names, which is more expensive for later hardforks.

I think we need something like this in order to support any type of hardfork.

what's not great is, is that this still requires to mixing ethereum hardforks with the network hardforks and doesn't treat ethereum hardforks as the base.

maybe we could find a way bake this in into the hardforks trait directly, this way we don't have a lot of overhead.

we can also use an additional map in the chainhardforks type for simpler lookups into the position in the array so we don't need to always iterate over all entries.